### PR TITLE
Ensure that $package_name is an array

### PR DESCRIPTION
**Pull request by Phil-Friderici**

### CODE REVIEW

**Comment by ghoneycutt** [comment]
package_name could be an array if you needed multiple packages. What about Solaris? Seems you need 3 packages to do anything there.

**Comment by anders-larsson** [comment]
@ghoneycutt I agree. That's why I did not include it while updating the README in #32.
If we decide to limit it to string only we should probably add verification that only strings are accepted.

Also, your commit includes the  cron.hourly support we spoke of, which does not reflect the commit message :)

**Comment by Phil-Friderici** [comment]
hmmmm.... I gues supporting an array for package_name would need some more rework on the dependencies. At the moment we don't support Solaris and all supported OSfamilies only have one package....
Let me try that with some spec tests ;)

**Comment by ghoneycutt** [comment]
I like the ability to have an array for any package because you might need multiple packages for local additions, such as `cron` and `ericsson_cron` where `ericsson_cron` might populate cron.d entries or whatever.

File: manifests/init.pp [hunk 1]
```diff
@@ -133,6 +133,7 @@
   validate_absolute_path($cron_weekly_path)
   validate_absolute_path($cron_monthly_path)
 
+  if is_string($package_name)     == false { fail('cron::package_name must be a string') }
```
Comment (ghoneycutt) [inline manifests/init.pp]: perhaps

``` puppet
if !is_string($package_name) and !is_array($package_name) { fail ('cron::package must be an array or a string') }
```

**Comment by Phil-Friderici** [comment]
Seems I shouldn't do drive heavy equipment today.
You both are right... it wasn't an issue to put an array into that.... will go to array instead of string.

**Comment by ghoneycutt** [comment]
Suggest that it supports both array and string

**Comment by Phil-Friderici** [comment]
this update will accept arrays and strings. Will convert a string to an array.
Documentation **intentionally** only talks about arrays.
@ghoneycutt @anders-larsson what do you think about this ?

File: README.md [hunk 2]
```diff
@@ -46,9 +46,9 @@ String for the ensure parameter for the cron package. Valid values are 'installe
 
 - *Default*: 'installed'
 
-package_name
-------------
-Name of the cron package. Specify this to change the platform's default.
+package_name (array)
```
Comment (anders-larsson) [inline README.md]: perhaps include use (string, array) to inform user it's OK with string as well.

Comment (Phil-Friderici) [inline README.md]: Documentation **intentionally** only talks about arrays.

File: manifests/init.pp [hunk 3]
```diff
@@ -78,9 +78,19 @@
...
-    $package_name_real = $package_name
+    case type3x($package_name) {
+      'Array': {
+        $package_name_array = $package_name
+      }
+      'String': {
+        $package_name_array = any2array($package_name)
+      }
+      default: {
+        fail("$package_name is not an Array.")
```
Comment (anders-larsson) [inline manifests/init.pp]: is not an Array or String?

File: README.md (hunk 2 above)
Comment (anders-larsson) [inline README.md]: ah. Cool with me :)

File: manifests/init.pp (hunk 3 above)
Comment (Phil-Friderici) [inline manifests/init.pp]: **intentionally** hahahaha

**Comment by anders-larsson** [comment]
Looks good to me in that case. Any reason why we wouldn't tell that strings are allowed? I myself prefer to use strings if there's only one entry because it looks wrong otherwise.

**Comment by Phil-Friderici** [comment]
Basic idea: getting the users used to choose the variable type that is finally used to configure the resources.

Is that too fishy ?

**Comment by ghoneycutt** [comment]
Clever, though I think that if you generally only have one package, then saying it must be declared as an array is a bit confusing. Least surprise principal would be to accept both (which you do) but to document both as well.

Saying that the documented way to change the package name in Hiera is 

``` yaml
cron::package_name:
  - my_cron
```

is a bit confusing since you would expect

``` yaml
cron::package_name: 'my_cron'
```

**Comment by Phil-Friderici** [comment]
"Least surprise principal" = good point....

Maybe it's because my parents were teaches ;)
Let me change the docs

**Comment by ghoneycutt** [comment]
Users shouldn't care about the data type other than using the one that works. ie: there shouldn't be a control mechanism like we do with foo_hiera_merge from the types module.

**Comment by Phil-Friderici** [comment]
updated the docs and error messages

**Comment by Phil-Friderici** [comment]
changed from require to before as it failed on Puppet v4

**Comment by Phil-Friderici** [comment]
RFC (request for comments)

**Pull request merged**